### ISSUE TITLE
Fix CasperJS tests

### ISSUE THREAD

**Issue by m-allanson** [MadeByLoosers/noise-box#31]
Fix CasperJS tests in the branch feature/casperjs.  See commit message 42d5f53ba072b17ead1549b4658f41f2818b9313 for more details.

**Comment by jedrichards** [comment]
Where you do `var casper = require('casper').create();` can you instead
make two instances of casper? Have one instance listening for events from
<audio> while the second instance plays a sound?

**Comment by m-allanson** [comment]
I tried something like that without much success, although that doesn't mean it wouldn't work.

From memory I tried creating two casper instances (imaginatively) titled `casper` and `casper2`.  In a `casper.then()` block I started running all the `casper2` tests, however I think `casper` was just carrying on to the next block while `casper2` was happily churning away, as `casper` had done everything it needed to.

Something a bit (but not exactly) like this: https://gist.github.com/4235225

**Comment by jedrichards** [comment]
Also, is it up possible to have the casper test file spin up a unique instance of the NoiseBox server just to test on, and then stop it when it's finished? It might involve a few changes to server.js, but I think it seems like a better approach in general?

That way you get to test that the server starts and stops properly, instead of just testing on one that's been manually started. Plus the tests will be operating on a different server/port to the dev version you just have running normally anyway and won't interfere with anything.

That's the way that James Shore was doing it on his TDD workshop, wasn't it? I only watched the first 7 or so though, do you know if he stuck with that approach?

**Comment by m-allanson** [comment]
I'm not sure if he stuck with it, but it makes sense.  You'd have to start the server with Grunt, as Casper has no awareness of that sort of thing.

I did try doing this before, but became unstuck when working out how to stop the app afterwards.  You can do `killall node` but that will clobber any other node instances that are running.  I also tried getting the app to send its process id to stdout when it starts, but couldn't work out how to get grunt to read that and use it later on. That would have allowed just killing that node instance with `kill pid`

One approach could be allowing the app to be started with a `testing` flag applied, which would then expose a route that would allow it to shut itself down.  Grunt could start the app, then the final test would be to check that visiting `/shutdown` would kill the app.  

Modifying the app to accommodate testing code doesn't feel like the cleanest thing to be doing.  But maybe it's ok?  Do you have any other suggestions on how to start and stop the app for the tests?

**Comment by jedrichards** [comment]
Maybe a bit of refactoring is required. Move `server.js` into `src/app` and have it export a few methods, perhaps something like this:

```
exports.start = function (port,onStartedCallback) { ... }
exports.stop = function () { ... }
```

Then instead of starting the app directly via `node server.js` start it instead via something like `node index.js`.

`index.js` would sit in our app's `src` root and `require("app/server")` and use its API to fire things up.

Similarly the casper script could `require("app/server")` and start it up in exactly the same way?

Basically refactor `server.js` so its more flexible and testable ... ?

**Comment by orangespaceman** [comment]
CasperJS doesn't officially support websockets.  Leaving this for now.
